In [54]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [55]:
import pandas as pd


In [56]:
file_path = "/content/drive/MyDrive/AnalystLab week 5/WA_Fn-UseC_-Telco-Customer-Churn.csv"

df = pd.read_csv(file_path)
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [57]:
df.shape

(7043, 21)

In [58]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [59]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [60]:
df['Churn'].value_counts()

,count
Churn,
No,5174
Yes,1869


In [61]:
df['Churn'].value_counts(normalize=True) * 100

,proportion
Churn,
No,73.463013
Yes,26.536987


In [62]:
df['TotalCharges'].unique()

array(['29.85', '1889.5', '108.15', ..., '346.45', '306.6', '6844.5'],
      dtype=object)

In [63]:
df['TotalCharges'].sort_values().head(20)

,TotalCharges
936,
3826,
4380,
753,
5218,
3331,
6754,
6670,
1340,
488,


In [64]:
df[df['TotalCharges'] == ' ']

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,Yes,Bank transfer (automatic),52.55,,No
753,3115-CZMZD,Male,0,No,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.25,,No
936,5709-LVOEQ,Female,0,Yes,Yes,0,Yes,No,DSL,Yes,...,Yes,No,Yes,Yes,Two year,No,Mailed check,80.85,,No
1082,4367-NUYAO,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.75,,No
1340,1371-DWPAZ,Female,0,Yes,Yes,0,No,No phone service,DSL,Yes,...,Yes,Yes,Yes,No,Two year,No,Credit card (automatic),56.05,,No
3331,7644-OMVMY,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,19.85,,No
3826,3213-VVOLG,Male,0,Yes,Yes,0,Yes,Yes,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,25.35,,No
4380,2520-SGTTA,Female,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,Two year,No,Mailed check,20.00,,No
5218,2923-ARZLG,Male,0,Yes,Yes,0,Yes,No,No,No internet service,...,No internet service,No internet service,No internet service,No internet service,One year,Yes,Mailed check,19.70,,No
6670,4075-WKNIU,Female,0,Yes,Yes,0,Yes,Yes,DSL,No,...,Yes,Yes,Yes,No,Two year,No,Mailed check,73.35,,No


In [65]:
df[df['TotalCharges'] == ' '][['customerID', 'tenure', 'MonthlyCharges', 'TotalCharges', 'Churn']]

,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,,No
753,3115-CZMZD,0,20.25,,No
936,5709-LVOEQ,0,80.85,,No
1082,4367-NUYAO,0,25.75,,No
1340,1371-DWPAZ,0,56.05,,No
3331,7644-OMVMY,0,19.85,,No
3826,3213-VVOLG,0,25.35,,No
4380,2520-SGTTA,0,20.00,,No
5218,2923-ARZLG,0,19.70,,No
6670,4075-WKNIU,0,73.35,,No


In [66]:
df['TotalCharges'] = df['TotalCharges'].replace(' ', 0)

In [67]:
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'])

In [68]:
df.duplicated().sum()

np.int64(0)

In [69]:
df.isnull().sum()

,0
customerID,0
gender,0
SeniorCitizen,0
Partner,0
Dependents,0
tenure,0
PhoneService,0
MultipleLines,0
InternetService,0
OnlineSecurity,0


In [70]:
for col in df.select_dtypes(include='object').columns:
  print(col)
  print(df[col].unique())
  print('-' * 50)

customerID
['7590-VHVEG' '5575-GNVDE' '3668-QPYBK' ... '4801-JZAZL' '8361-LTMKD'
 '3186-AJIEK']
--------------------------------------------------
gender
['Female' 'Male']
--------------------------------------------------
Partner
['Yes' 'No']
--------------------------------------------------
Dependents
['No' 'Yes']
--------------------------------------------------
PhoneService
['No' 'Yes']
--------------------------------------------------
MultipleLines
['No phone service' 'No' 'Yes']
--------------------------------------------------
InternetService
['DSL' 'Fiber optic' 'No']
--------------------------------------------------
OnlineSecurity
['No' 'Yes' 'No internet service']
--------------------------------------------------
OnlineBackup
['Yes' 'No' 'No internet service']
--------------------------------------------------
DeviceProtection
['No' 'Yes' 'No internet service']
--------------------------------------------------
TechSupport
['No' 'Yes' 'No internet service']
------------

### Data Cleaning Summary

- The dataset contains 7,043 rows and 21 columns.
- No duplicate records were found.
- No missing values were found after inspection.
- The `TotalCharges` column was originally stored as object/text because it contained blank spaces.
- 11 blank `TotalCharges` records were found, all with `tenure = 0`, indicating likely new customers.
- The blank values were replaced with 0 and `TotalCharges` was converted to float.
- Categorical columns were checked and appeared consistent.

In [71]:
cleaned_path = "/content/drive/MyDrive/AnalystLab week 5/WA_Fn-UseC_-Telco-Customer-Churn-Cleaned.csv"
df.to_csv(cleaned_path, index=False)

**Exploratory Data Analysis**

In [72]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges
count,7043.000000,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692,2279.734304
std,0.368612,24.559481,30.090047,2266.794470
min,0.000000,0.000000,18.250000,0.000000
25%,0.000000,9.000000,35.500000,398.550000
50%,0.000000,29.000000,70.350000,1394.550000
75%,0.000000,55.000000,89.850000,3786.600000
max,1.000000,72.000000,118.750000,8684.800000


In [73]:
df['Churn'].value_counts(normalize=True) * 100

,proportion
Churn,
No,73.463013
Yes,26.536987


In [74]:
pd.crosstab(df['Contract'], df['Churn'], normalize='index') * 100

Churn,No,Yes
Contract,,
Month-to-month,57.290323,42.709677
One year,88.730482,11.269518
Two year,97.168142,2.831858


In [75]:
df['TenureGroup'] = pd.cut(
    df['tenure'],
    bins=[-1, 12, 24, 48, 72],
    labels=['0-12 months', '13-24 months', '25-48 months', '49-72 months']
    )


In [76]:
pd.crosstab(df['TenureGroup'], df['Churn'], normalize='index') * 100

Churn,No,Yes
TenureGroup,,
0-12 months,52.561757,47.438243
13-24 months,71.289062,28.710938
25-48 months,79.611041,20.388959
49-72 months,90.486824,9.513176


In [77]:
df.groupby('Churn')['MonthlyCharges'].mean()

,MonthlyCharges
Churn,
No,61.265124
Yes,74.441332


In [78]:
df['MonthlyChargeGroup'] = pd.cut(
    df['MonthlyCharges'],
    bins=[0, 35, 70, 120],
    labels=['Low charge', 'Medium charge', 'High charge']
)

pd.crosstab(df['MonthlyChargeGroup'], df['Churn'], normalize='index') * 100

Churn,No,Yes
MonthlyChargeGroup,,
Low charge,89.106628,10.893372
Medium charge,76.057971,23.942029
High charge,64.638571,35.361429


In [79]:
pd.crosstab(df['InternetService'], df['Churn'], normalize='index') * 100

Churn,No,Yes
InternetService,,
DSL,81.040892,18.959108
Fiber optic,58.107235,41.892765
No,92.595020,7.404980


In [80]:
df.groupby('InternetService')['MonthlyCharges'].mean()

,MonthlyCharges
InternetService,
DSL,58.102169
Fiber optic,91.500129
No,21.079194


In [81]:
pd.crosstab(df['TechSupport'], df['Churn'], normalize='index') * 100

Churn,No,Yes
TechSupport,,
No,58.364526,41.635474
No internet service,92.595020,7.404980
Yes,84.833659,15.166341


In [82]:
pd.crosstab(df['PaymentMethod'], df['Churn'], normalize='index') * 100

Churn,No,Yes
PaymentMethod,,
Bank transfer (automatic),83.290155,16.709845
Credit card (automatic),84.756899,15.243101
Electronic check,54.714588,45.285412
Mailed check,80.893300,19.106700


In [83]:
pd.crosstab(df['DeviceProtection'], df['Churn'], normalize='index') * 100

Churn,No,Yes
DeviceProtection,,
No,60.872375,39.127625
No internet service,92.595020,7.404980
Yes,77.497936,22.502064


In [31]:
pd.crosstab(df['OnlineSecurity'], df['Churn'], normalize='index') * 100

Churn,No,Yes
OnlineSecurity,,
No,58.233276,41.766724
No internet service,92.595020,7.404980
Yes,85.388806,14.611194


In [32]:
pd.crosstab(df['PaperlessBilling'], df['Churn'], normalize='index') * 100

Churn,No,Yes
PaperlessBilling,,
No,83.669916,16.330084
Yes,66.434908,33.565092


In [33]:
pd.crosstab(df['SeniorCitizen'], df['Churn'], normalize='index') * 100

Churn,No,Yes
SeniorCitizen,,
0,76.393832,23.606168
1,58.318739,41.681261


## Final Summary Tables and Export

In [84]:
df['Churn'].unique()

array(['No', 'Yes'], dtype=object)

In [85]:
if df['Churn'].dtype == 'object':
    df['ChurnFlag'] = df['Churn'].str.strip().map({'Yes': 1, 'No': 0})
else:
    df['ChurnFlag'] = df['Churn'].astype(int)

# Confirm ChurnFlag is correct
df[['Churn', 'ChurnFlag']].head()

,Churn,ChurnFlag
0,No,0
1,No,0
2,Yes,1
3,No,0
4,Yes,1


In [86]:
df['ChurnFlag'].isnull().sum()

np.int64(0)

In [87]:
df['ChurnFlag'].value_counts()

,count
ChurnFlag,
0,5174
1,1869


In [88]:
# Create tenure groups
df['TenureGroup'] = pd.cut(
    df['tenure'],
    bins=[-1, 12, 24, 48, 72],
    labels=['0-12 months', '13-24 months', '25-48 months', '49-72 months']
)

# Create monthly charge groups
df['MonthlyChargeGroup'] = pd.cut(
    df['MonthlyCharges'],
    bins=[0, 35, 70, 120],
    labels=['Low charge', 'Medium charge', 'High charge']
)

df[['tenure', 'TenureGroup', 'MonthlyCharges', 'MonthlyChargeGroup']].head()

,tenure,TenureGroup,MonthlyCharges,MonthlyChargeGroup
0,1,0-12 months,29.85,Low charge
1,34,25-48 months,56.95,Medium charge
2,2,0-12 months,53.85,Medium charge
3,45,25-48 months,42.30,Medium charge
4,2,0-12 months,70.70,High charge


In [89]:
def churn_summary(column):
    summary = df.groupby(column, observed=False).agg(
        Total_Customers=('ChurnFlag', 'count'),
        Churned_Customers=('ChurnFlag', 'sum')
    ).reset_index()

    summary['Churn_Rate'] = (
        summary['Churned_Customers'] / summary['Total_Customers']
    ) * 100

    summary['Churn_Rate'] = summary['Churn_Rate'].round(2)

    return summary.sort_values(by='Churn_Rate', ascending=False)

In [90]:
overall_churn_summary = df['ChurnFlag'].value_counts().reset_index()
overall_churn_summary.columns = ['ChurnFlag', 'Customer_Count']

overall_churn_summary['Churn_Status'] = overall_churn_summary['ChurnFlag'].map({
    0: 'No Churn',
    1: 'Churn'
})

overall_churn_summary['Percentage'] = (
    overall_churn_summary['Customer_Count'] / len(df)
) * 100

overall_churn_summary['Percentage'] = overall_churn_summary['Percentage'].round(2)

overall_churn_summary = overall_churn_summary[
    ['Churn_Status', 'Customer_Count', 'Percentage']
]

In [91]:
contract_summary = churn_summary('Contract')
tenure_summary = churn_summary('TenureGroup')
monthly_charge_summary = churn_summary('MonthlyChargeGroup')
internet_service_summary = churn_summary('InternetService')
payment_method_summary = churn_summary('PaymentMethod')

In [92]:
overall_churn_summary

,Churn_Status,Customer_Count,Percentage
0,No Churn,5174,73.46
1,Churn,1869,26.54


In [93]:
contract_summary

,Contract,Total_Customers,Churned_Customers,Churn_Rate
0,Month-to-month,3875,1655,42.71
1,One year,1473,166,11.27
2,Two year,1695,48,2.83


In [94]:
tenure_summary

,TenureGroup,Total_Customers,Churned_Customers,Churn_Rate
0,0-12 months,2186,1037,47.44
1,13-24 months,1024,294,28.71
2,25-48 months,1594,325,20.39
3,49-72 months,2239,213,9.51


In [95]:
monthly_charge_summary

,MonthlyChargeGroup,Total_Customers,Churned_Customers,Churn_Rate
2,High charge,3583,1267,35.36
1,Medium charge,1725,413,23.94
0,Low charge,1735,189,10.89


In [96]:
internet_service_summary

,InternetService,Total_Customers,Churned_Customers,Churn_Rate
1,Fiber optic,3096,1297,41.89
0,DSL,2421,459,18.96
2,No,1526,113,7.40


In [97]:
payment_method_summary

,PaymentMethod,Total_Customers,Churned_Customers,Churn_Rate
2,Electronic check,2365,1071,45.29
3,Mailed check,1612,308,19.11
0,Bank transfer (automatic),1544,258,16.71
1,Credit card (automatic),1522,232,15.24


In [98]:
output_folder = "/content/drive/MyDrive/AnalystLab week 5"

In [99]:
df.to_csv(f"{output_folder}/WA_Fn-UseC_-Telco-Customer-Churn-Cleaned.csv", index=False)

overall_churn_summary.to_csv(f"{output_folder}/overall_churn_summary.csv", index=False)
contract_summary.to_csv(f"{output_folder}/contract_churn_summary.csv", index=False)
tenure_summary.to_csv(f"{output_folder}/tenure_churn_summary.csv", index=False)
monthly_charge_summary.to_csv(f"{output_folder}/monthly_charge_churn_summary.csv", index=False)
internet_service_summary.to_csv(f"{output_folder}/internet_service_churn_summary.csv", index=False)
payment_method_summary.to_csv(f"{output_folder}/payment_method_churn_summary.csv", index=False)

In [100]:
import os
os.listdir(output_folder)

['WA_Fn-UseC_-Telco-Customer-Churn.csv',
 'overall_churn_summary.csv',
 'contract_churn_summary.csv',
 'tenure_churn_summary.csv',
 'monthly_charge_churn_summary.csv',
 'internet_service_churn_summary.csv',
 'payment_method_churn_summary.csv',
 'WA_Fn-UseC_-Telco-Customer-Churn-Cleaned.csv']

In [101]:
df.to_csv(f"{output_folder}/WA_Fn-UseC_-Telco-Customer-Churn-Cleaned.csv", index=False)